src/utilities/HiFiMath.ts

```diff
@@ -0,0 +1,464 @@
+            axis.z = this.z / imaginaryLength;
+            let wholeLength = Math.sqrt(imaginaryLength + this.w * this.w);
+            let angle = 2.0 * Math.acos(this.w / wholeLength);
+            if (angle < 0.0) {
+                // we choose the axis that corresponds to positive angle
+                axis.negate();
+            }
+        }
+        return axis;
+    }
+
+    /**
+     * @returns Euler angle decomposition object: {yaw:, pitch:, roll:}
+     * where:
+     *   yaw = degrees rotation about 'up' axis
+     *   pitch = degrees rotation about yawed 'right' axis
+     *   roll = degrees rotation about yawed and pitched 'forward' axis
+     */
+    getEulerAngles() {
+        let forward = new Vector3({x: 0.0, y: 0.0, z: -1.0});
+        let rotatedForward = this.rotateVector(forward);
+
+        let yaw = 0.0;
+        let projectedLengthSquared = rotatedForward.x * rotatedForward.x + rotatedForward.z * rotatedForward.z;
+        let right = new Vector3({x: 1.0, y: 0.0, z: 0.0});
+        if (projectedLengthSquared > MIN_NORMALIZABLE_SQUARE_LENGTH) {
+            // rotatedForward has a non-zero component in the horizontal plane
+            // and we use that projection to compute the angle
+            yaw = Math.acos(Vector3.dot(forward, rotatedForward) / Math.sqrt(projectedLengthSquared)); 
+            if (Vector3.dot(rotatedForward, right) > 0.0) {
+                yaw *= -1;
+            }
+        } else {
+            // rotatedForward points along vertical axis and has no projection on the horizontal plane
+            // however the "rotatedRight" axis effectively lies on the horizontal plane
+            // and we can extract the azimuthal angle from its projection there
+            let rotatedRight = this.rotateVector(right);
+            projectedLengthSquared = rotatedRight.x * rotatedRight.x + rotatedRight.z * rotatedRight.z;
+            yaw = Math.acos(Vector3.dot(rotatedRight, right) / Math.sqrt(projectedLengthSquared)); 
+            if (Vector3.dot(rotatedRight, forward) < 0.0) {
+                yaw *= -1;
+            }
+        }
+
+        let pitch = Math.asin(rotatedForward.y);
+
+        // For "roll": rotatedUp direction lives on a plane spanned by two unit vectors: yawedRight and pitchedUp
+        // We can compute "roll" using trigonometry once we decompose rotatedUp onto that plane
+        let up = new Vector3({ x: 0.0, y: 1.0, z: 0.0 });
+        let rotatedUp = this.rotateVector(up);
+        let yawedRight = new Vector3({ x: Math.cos(yaw), y: 0.0, z: -Math.sin(yaw) });
+        let pitchedUp = Vector3.cross(yawedRight, rotatedForward); // note, this is already unitary
+        let roll = Math.atan2( Vector3.dot(rotatedUp, yawedRight), Vector3.dot(rotatedUp, pitchedUp));
+
+        // convert all angles to degrees
+        yaw *= RADIANS_TO_DEGREES;
+        pitch *= RADIANS_TO_DEGREES;
+        roll *= RADIANS_TO_DEGREES;
+
+        return {"yaw": yaw, "pitch": pitch, "roll": roll};
```

Comment:
Here (and potentially elsewhere similar), please change the key names for this object to include units, i.e.:

```js
        return {"yawDegrees": yaw, "pitchDegrees": pitch, "rollDegrees": roll};
```